src/charm.py

```diff
@@ -181,6 +181,94 @@ def _on_synapse_nginx_pebble_ready(self, _: ops.HookEvent) -> None:
+        """Get peer relation.
+
+        Returns:
+            Relation or not if is not found.
+        """
+        peer_relation = self.model.get_relation(PEER_RELATION_NAME)
+        return peer_relation
+
+    def get_admin_access_token(self) -> typing.Optional[str]:
+        """Get admin access token.
+
+        Returns:
+            admin access token or None if fails.
+        """
+        peer_relation = self._get_peer_relation()
+        if not peer_relation:
+            logger.error(
+                "Failed to get admin access token: no peer relation %s found", PEER_RELATION_NAME
+            )
+            return None
+        admin_access_token = None
+        if JUJU_HAS_SECRETS:
+            secret_id = peer_relation.data[self.app].get(SECRET_ID)
+            if secret_id:
+                try:
+                    secret = self.model.get_secret(id=secret_id)
+                    admin_access_token = secret.get_content().get(SECRET_KEY)
+                    return admin_access_token
+                except ops.model.SecretNotFoundError as exc:
+                    logger.exception("Failed to get secret id %s: %s", secret_id, str(exc))
+                    del peer_relation.data[self.app][SECRET_ID]
+                    return None
+        else:
+            # There is no Secrets support and none relation data was created
+            # So lets create the user and store its token in the peer relation
+            secret_value = peer_relation.data[self.app].get(SECRET_KEY)
+            if secret_value:
+                return secret_value
+        return None
+
+    def _start_synapse_stats_exporter(self) -> None:
+        """Start Synapse Stats Exporter in Synapse container."""
+        container = self.unit.get_container(synapse.SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            logger.error("Failed to connect to container while starting Synapse Stats Exporter")
+            return
+        self.pebble_service.replan_stats_exporter(container)
+
+    def _set_admin_access_token(self, access_token: str) -> None:
+        """Set admin access token.
+
+        Args:
+            access_token: token to save in the secret.
+        """
+        peer_relation = self._get_peer_relation()
+        if not peer_relation:
+            logger.error(
+                "Failed to get admin access token: no peer relation %s found", PEER_RELATION_NAME
+            )
+            return
+        if JUJU_HAS_SECRETS:
+            logger.debug("Adding secret")
+            secret = self.app.add_secret({SECRET_KEY: access_token})
+            peer_relation.data[self.app].update({SECRET_ID: secret.id})
+        else:
+            logger.debug("Adding peer data")
+            peer_relation.data[self.app].update({SECRET_KEY: access_token})
+
+    def _on_start(self, _: ops.HookEvent) -> None:
+        """Handle start event."""
+        if self.get_admin_access_token():
+            self._start_synapse_stats_exporter()
```

Comment:
Doesn'g mjolnir require an access token as well? Should it be added here?

Reply:
Yes, Mjolnir requires it too.

- If Mjolnir is enabled, the charm will instantiate a Mjolnir class passing state as parameter
- The class observes the collect status event that will check for the requirements and enable Mjolnir when everything is set